### PR TITLE
Adding HSDP-1D to the Device List

### PR DESCRIPTION
**Pull request by Vaasref**
The simplest and less expensive (by no mean cheap) printer from the company specialized in e-warfare, SlapDragon.

The speed is not defined yet. But preliminary concept was 0.5 page/s with 40 lines/page.

The Device ID ans Device Type are also not defined since there is no corresponding type yet.

### CODE REVIEW

**Comment by andyleap** [comment]
20 lines per second is crazy fast.  I would think something like 1~2 lines a second would be good.  There should also be some mention of the fact that a spark printer will require specialized paper, most likely black paper with a very thin layer of aluminum.

**Comment by sctjkc01** [comment]
You might wish to consider putting the HSDP in the device list (TC-IDs.md) as well...

Additionally, just because a type doesn't exist didn't stop others from devising their own subtypes.  You can make your own...

**Comment by Vaasref** [comment]
@andyleap : The purpose of the printer is to dump data.

It needs to be fast to be able to do real time data logging.
If it can't be fast enough then it can't be SlapDragon since SlapDragon would not design a product for something else than e-war.

So if you are talking about black paper, I assume you looked at printing technologies. then you would also have seen that 20 lines/s is realistic.

Of course using sparks to print will come with high power consumption and EM sig.

**Comment by andyleap** [comment]
No, 20 lines/s is not realistic.   Not for in game, just barely for real life.  If you were talking impact printer, you might have a chance, but while a high speed modern line printer could do up to 2000 lines per minute, putting that kind of thing into the game is insane.  Not to mention that every example of spark printers that I've seen is sitting at a line per second or so.  If you are wanting to log data, I would suggest going with something that's actually meant to be a storage device.

**Comment by Vaasref** [comment]
What is sheet of paper if not a storage device ?

With the storage capacity of current HDD you can't beat the paper.
With 3MiB for every 1000pages in text mode.

And paper is a pain to hack into btw. Good luck trying to destroy records remotely, but if you want to access data you need a miracle.

I can still revert to simple line matrix if that really bugs you.

Also to talk more about how to make it in game, a block of paper is dead simple to model.
The printer itself only has one moving part (a rotating cylinder) and is also visually pretty boring and simple.
I don't understand why you say that 20l/s is "insane" for a game.

**Comment by andyleap** [comment]
Paper is not a storage device, it's a static output device.  And I said impact, not line printer, as an impact printer does match up with what you are describing, and can be very fast..  Plus, what's the point of outputting 20 lines per second when you can't even read that fast?  Are you planning on putting a scanner in the game too?

Edit: another thing, go have fun hacking a floppy disk sitting on a desk remotely.

**Comment by Vaasref** [comment]
The point of logging is logging. You don't need to read logs as fast as it come.
If you have a sensor that does not work with magic, your need to log things.

And with a printer you don't need to manually remove te paper from it to make it hack safe.

You do with a floppies.
Also floppies are vulnerable to dust, heat, EM attacks, solar flares and they cannot be read by humans.
Floppies are a poor choice for security.

Paper is a storage media, when you write something on paper you are storing data on it. It's the whole point of paper.

And for that kind of purpose it's the best.

**Comment by sctjkc01** [comment]
If we’re listing the fallacies of floppies, might I just point out that paper...
- Is not too good with heat (too much and it’ll ignite).
  - Is a fire hazard.
- Is not too good with water (too much and it’s soggy, turning into a mushy pulp).
- Is insecure to attacks with the visible spectrum (can be read by literally _anyone_).
- Requires light to read (cannot be read in the dark).
- Requires a human to transfer information back into a computer.
- Can only be used once (print something on it, and that surface can't be used again).

**_NO_** storage medium (using the loosest form of the term) is without its cons.

**Comment by paultech** [comment]
This got a bit out of hand

Changes needed to merge:
TC-IDs.md needs updated with a printer type
The speed of the printer needs to be decreased 5 lines/second is in range for me.

File: HSDP-1D.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Manufacturer ID   : 0xF6976D00
+
+-----------------------------------------------------------------------------------------
+Description :
+-----------------------------------------------------------------------------------------
+The HSDP line is all about printing as fast as possible the data you need.
+
+The HSDP offer you the best way to store you data in term of cost efficency.
+Paper power and data is all you need to run a HSDP, its clever design combining the 
+speed of the line matrix and the lack of nedd for ink of a spark printer design will
```
Comment (andyleap) [inline HSDP-1D.md:41]: "lack of nedd for ink" need*

File: HSDP-1D.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+-----------------------------------------------------------------------------------------
+The HSDP line is all about printing as fast as possible the data you need.
+
+The HSDP offer you the best way to store you data in term of cost efficency.
+Paper power and data is all you need to run a HSDP, its clever design combining the 
+speed of the line matrix and the lack of nedd for ink of a spark printer design will 
+really cut the cost of your radio survey by orders of magnitude.
+
+The HSDP line is also very quiet, SlapDragon guarantee less than 95dB of noise at full 
+speed !
```
Comment (andyleap) [inline HSDP-1D.md:45]: Space between word and punctuation

File: HSDP-1D.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+The HSDP-1D is a Data version of the model 1, equiped with curring edge memory chips
```
Comment (andyleap) [inline HSDP-1D.md:47]: "equiped with curring edge memory" equipped\* cutting*

File: HSDP-1D.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+The HSDP offer you the best way to store you data in term of cost efficency.
+Paper power and data is all you need to run a HSDP, its clever design combining the 
+speed of the line matrix and the lack of nedd for ink of a spark printer design will 
+really cut the cost of your radio survey by orders of magnitude.
+
+The HSDP line is also very quiet, SlapDragon guarantee less than 95dB of noise at full 
+speed !
+
+The HSDP-1D is a Data version of the model 1, equiped with curring edge memory chips 
+and buses this printer will be able to log everything you throw at it.
```
Comment (andyleap) [inline HSDP-1D.md:48]: "and buses this" busses*

File: HSDP-1D.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+It is built to print large amount of data in a smaller amount of time and lack 
+typesettings possibilities of the Graph variant.
```
Comment (andyleap) [inline HSDP-1D.md:50]: "and lack typesettings possibilities" -> "and lacks the typesetting possibilities"

File: HSDP-1D.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+3 : Print simple line
+    Reads register B and prints a full line (according to current mode) from DCPU's 
+	memory stating at B
+
+3 : Print multiple lines
+    Reads register B and C and prints C full lines (according to current mode) from 
+	DCPU's memory stating at B
+ 
+4 : Full Dump
+    Reads and prints the full memory of the DCPU halting
```
Comment (andyleap) [inline HSDP-1D.md:81]: halting what?  halting the DCPU while the dump is running?

File: HSDP-1D.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+3 : Print multiple lines
+    Reads register B and C and prints C full lines (according to current mode) from 
+	DCPU's memory stating at B
+ 
+4 : Full Dump
+    Reads and prints the full memory of the DCPU halting 
+
+
+! WARNING !
+Printing more than 2048 bytes of data may in 1 interupt result in halting your DCPU.
```
Comment (andyleap) [inline HSDP-1D.md:85]: "Printing more than 2048 bytes of data may in 1 interupt result in halting your DCPU." -> "Printing more than 2048 bytes of data in one interrupt may result in the DCPU halting until the printer can catch up."  May want to clear this up some, what about 2 back to back interrupts 1ms apart with 2048 bytes apiece?  Why isn't this buffer length listed anywhere else?

File: HSDP-1D.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+0 : Set Mode
+    Reads register B and set the mode at B&0x1
+	
+1 : Get Mode
+    Set DCPU register A to current mode
+	
+2 : Cut page
+    Cuts the last printed page
+
+3 : Print simple line
```
Comment (andyleap) [inline HSDP-1D.md:72]: simple or single?

File: HSDP-1D.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
+```
```
Comment (andyleap) [inline HSDP-1D.md:1]: Is there some reason this entire spec is a code block in markdown?  Makes it miss out on the formatting that could be happening.

File: HSDP-1D.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+Device Name       : High Speed Data Printer Model 1Data (HSDP-1D)
+Device ID         : TBD
+Device Type       : TBD
+Manufacturer Name : SlapDragon
+Manufacturer ID   : 0xF6976D00
+
+-----------------------------------------------------------------------------------------
+Description :
+-----------------------------------------------------------------------------------------
+The HSDP line is all about printing as fast as possible the data you need.
```
Comment (andyleap) [inline HSDP-1D.md:37]: Phrasing awkward? maybe "The HSDP line is all about printing the data you need as fast as possible ."?

File: HSDP-1D.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+The HSDP offer you the best way to store you data in term of cost efficency.
```
Comment (andyleap) [inline HSDP-1D.md:39]: "store you data in term of cost efficency" -> "store your data in terms of cost efficiency"

File: HSDP-1D.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+Paper power and data is all you need to run a HSDP, its clever design combining the
```
Comment (andyleap) [inline HSDP-1D.md:40]: "Paper, power, and data are all you need" "it's"

File: HSDP-1D.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    Reads register B and C and prints C full lines (according to current mode) from 
+	DCPU's memory stating at B
+ 
+4 : Full Dump
+    Reads and prints the full memory of the DCPU halting 
+
+
+! WARNING !
+Printing more than 2048 bytes of data may in 1 interupt result in halting your DCPU.
+But be reassured that no printer of the HSDP line can suffer from any amount of data
```
Comment (andyleap) [inline HSDP-1D.md:86]: "But be assured"

**Review by andyleap — CHANGES_REQUESTED** [review]
Also need to insert the printing speed into the spec as well.

**Comment by paultech** [comment]
superseded by #34

**Pull request closed without merging**